### PR TITLE
Aot watch

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/compiler-cli/src/ng_host.ts [hunk 1]
```diff
@@ -75,7 +76,7 @@ export class NgHost implements AotCompilerHost {
    *
    * NOTE: (*) the relative path is computed depending on `isGenDirChildOfRootDir`.
    */
-  getImportPath(containingFile: string, importedFile: string): string {
+  resolveFileToImport(importedFile: string, containingFile: string): string {
```
Comment (alexeagle) [inline modules/@angular/compiler-cli/src/ng_host.ts]: fileToRelativeImport?
https://nodejs.org/api/path.html#path_path_relative_from_to

I would prefer these APIs match my understanding from Node's path handling

File: modules/@angular/compiler-cli/test/ng_host_spec.ts [hunk 2]
```diff
@@ -57,87 +56,87 @@ describe('NgHost', () => {
...
-                 '/tmp/project/src/my.ngfactory.ts', '/tmp/project/src/a/my.other.css.shim.ts'))
+      expect(hostNestedGenDir.resolveFileToImport(
+                 '/tmp/project/src/a/my.other.css.shim.ts', '/tmp/project/src/my.ngfactory.ts'))
           .toEqual('./a/my.other.css.shim');
     });
 
     it('should import application from factory', () => {
-      expect(hostNestedGenDir.getImportPath(
-                 '/tmp/project/src/my.ngfactory.ts', '/tmp/project/src/my.other.ts'))
+      expect(hostNestedGenDir.resolveFileToImport(
```
Comment (alexeagle) [inline modules/@angular/compiler-cli/test/ng_host_spec.ts]: I'd like the order to match node: from, to

File: tools/@angular/tsc-wrapped/package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@angular/tsc-wrapped",
-  "version": "0.2.2",
+  "version": "0.4.1",
```
Comment (alexeagle) [inline tools/@angular/tsc-wrapped/package.json]: 0.4.0 still until release

File: modules/@angular/compiler-cli/src/ng_host.ts [hunk 4]
```diff
@@ -150,43 +150,69 @@ export class NgHost implements AotCompilerHost {
...
+        v2Metadata = {'__symbolic': 'module', 'version': 2, 'metadata': {}};
+        if (v1Metadata.exports) {
+          v2Metadata.exports = v1Metadata.exports;
+        }
+        for (let prop in v1Metadata.metadata) {
+          v2Metadata.metadata[prop] = v1Metadata.metadata[prop];
+        }
+        const sourceText = this.context.readFile(dtsFilePath);
+        const exports = this.metadataCollector.getMetadata(
+            ts.createSourceFile(dtsFilePath, sourceText, ts.ScriptTarget.Latest, true));
```
Comment (alexeagle) [inline modules/@angular/compiler-cli/src/ng_host.ts]: first look for it in the program?

**Review by alexeagle — APPROVED** [review]

File: modules/@angular/compiler-cli/src/ng_host.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/compiler-cli/src/ng_host.ts]: Done

File: tools/@angular/tsc-wrapped/package.json (hunk 3 above)
Comment (tbosch) [inline tools/@angular/tsc-wrapped/package.json]: Done

File: modules/@angular/compiler/src/aot/compiler_host.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+   * produced and the module has exported variables or classes with decorators. Module metadata can
+   * also be produced directly from TypeScript sources by using MetadataCollector in tools/metadata.
+   *
+   * @param modulePath is a string identifier for a module as an absolute path.
+   * @returns the metadata for the given module.
+   */
+  getMetadataFor(modulePath: string): {[key: string]: any}|{[key: string]: any}[];
+
+  /**
+   * Converts a module name into a file path.
```
Comment (tbosch) [inline modules/@angular/compiler/src/aot/compiler_host.ts]: Added.

File: modules/@angular/compiler-cli/src/ng_host.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler-cli/src/ng_host.ts]: Changed to `moduleNameToFileName` and `fileNameToModuleName`.

File: modules/@angular/compiler-cli/src/codegen.ts [hunk 6]
```diff
@@ -17,9 +17,9 @@ import {readFileSync} from 'fs';
 import * as path from 'path';
 import * as ts from 'typescript';
 
-import {PathMappedReflectorHost} from './path_mapped_reflector_host';
+import {NgHost, NgHostContext} from './ng_host';
```
Comment (tbosch) [inline modules/@angular/compiler-cli/src/codegen.ts]: Done

**Comment by tbosch** [comment]
Needs to be synced into G3 with cl/139588792

**Comment by tbosch** [comment]
Reviewed in person with @alexeagle

**Pull request merged**